File: ios-amazon-clone/NovaMart/Stores/PaymentStore.swift
import Foundation
import Observation

@Observable
final class PaymentStore {
    var methods: [SavedPaymentMethod] = []
    var defaultMethodId: String?

    init() {
        self.methods = PersistenceService.load([SavedPaymentMethod].self, forKey: PersistenceService.Keys.paymentMethods) ?? []
        self.defaultMethodId = UserDefaults.standard.string(forKey: PersistenceService.Keys.defaultPaymentId)
    }

    var defaultMethod: SavedPaymentMethod? {
        methods.first { $0.id == defaultMethodId }
    }

    func addMethod(_ type: PaymentMethodType, label: String, cardNumber: String? = nil, nameOnCard: String? = nil, expiry: String? = nil, upiId: String? = nil, bankName: String? = nil, setAsDefault: Bool = false) {
        let id = Helpers.generateId()
        let isFirst = methods.isEmpty
        let shouldBeDefault = setAsDefault || isFirst

        if shouldBeDefault {
            for i in methods.indices {
                methods[i].isDefault = false
            }
        }

        var cardLast4: String?
        var maskedNumber: String?
        if let cardNumber {
            cardLast4 = Formatters.cardLast4(cardNumber)
            maskedNumber = Formatters.maskCardNumber(cardNumber)
        }

        let method = SavedPaymentMethod(
            id: id,
            type: type,
            label: label,
            isDefault: shouldBeDefault,
            cardNumber: maskedNumber,
            cardLast4: cardLast4,
            nameOnCard: nameOnCard,
            expiry: expiry,
            upiId: upiId,
            bankName: bankName
        )
        methods.append(method)

        if shouldBeDefault {
            defaultMethodId = id
        }
        persist()
    }

    func removeMethod(_ id: String) {
        let wasDefault = defaultMethodId == id
        methods.removeAll { $0.id == id }

        if wasDefault {
            let newDefaultId = methods.first?.id
            defaultMethodId = newDefaultId
            for i in methods.indices {
                methods[i].isDefault = methods[i].id == newDefaultId
            }
        }
        persist()
    }

    func setDefault(_ id: String) {
        defaultMethodId = id
        for i in methods.indices {
            methods[i].isDefault = methods[i].id == id
        }
        persist()
    }

    private func persist() {
        PersistenceService.save(methods, forKey: PersistenceService.Keys.paymentMethods)
        if let defaultMethodId {
            UserDefaults.standard.set(defaultMethodId, forKey: PersistenceService.Keys.defaultPaymentId)
        } else {
            UserDefaults.standard.removeObject(forKey: PersistenceService.Keys.defaultPaymentId)
        }
    }
}

File: amazon-clone/src/store/paymentStore.ts
import { create } from "zustand";
import { persist } from "zustand/middleware";
import { SavedPaymentMethod, PaymentMethodType } from "@/lib/types";
import { generateId } from "@/lib/utils";

interface PaymentState {
  methods: SavedPaymentMethod[];
  defaultMethodId: string | null;
  addMethod: (method: Omit<SavedPaymentMethod, "id" | "isDefault">, setAsDefault?: boolean) => void;
  updateMethod: (id: string, updates: Partial<Omit<SavedPaymentMethod, "id" | "isDefault">>) => void;
  removeMethod: (id: string) => void;
  setDefault: (id: string) => void;
  getDefault: () => SavedPaymentMethod | undefined;
}

export const usePaymentStore = create<PaymentState>()(
  persist(
    (set, get) => ({
      methods: [],
      defaultMethodId: null,

      addMethod: (method, setAsDefault = false) => {
        const id = generateId();
        const isFirst = get().methods.length === 0;
        const isDefault = setAsDefault || isFirst;

        set((state) => ({
          methods: [
            ...state.methods.map((m) =>
              isDefault ? { ...m, isDefault: false } : m
            ),
            { ...method, id, isDefault },
          ],
          defaultMethodId: isDefault ? id : state.defaultMethodId,
        }));
      },

      updateMethod: (id, updates) => {
        set((state) => ({
          methods: state.methods.map((m) =>
            m.id === id ? { ...m, ...updates } : m
          ),
        }));
      },

      removeMethod: (id) => {
        set((state) => {
          const remaining = state.methods.filter((m) => m.id !== id);
          const wasDefault = state.defaultMethodId === id;
          const newDefaultId = wasDefault
            ? remaining[0]?.id ?? null
            : state.defaultMethodId;

          return {
            methods: remaining.map((m) => ({
              ...m,
              isDefault: m.id === newDefaultId,
            })),
            defaultMethodId: newDefaultId,
          };
        });
      },

      setDefault: (id) => {
        set((state) => ({
          methods: state.methods.map((m) => ({
            ...m,
            isDefault: m.id === id,
          })),
          defaultMethodId: id,
        }));
      },

      getDefault: () => {
        const state = get();
        return state.methods.find((m) => m.id === state.defaultMethodId);
      },
    }),
    {
      name: "novamart-payments",
    }
  )
);

export function getPaymentDisplayText(method: SavedPaymentMethod): string {
  switch (method.type) {
    case "credit_card":
      return `Credit Card ending in ${method.cardLast4}`;
    case "debit_card":
      return `Debit Card ending in ${method.cardLast4}`;
    case "upi":
      return `UPI - ${method.upiId}`;
    case "net_banking":
      return `Net Banking - ${method.bankName}`;
    default:
      return method.label;
  }
}

export function getPaymentMethodLabel(type: PaymentMethodType): string {
  switch (type) {
    case "credit_card":
      return "Credit Card";
    case "debit_card":
      return "Debit Card";
    case "upi":
      return "UPI";
    case "net_banking":
      return "Net Banking";
  }
}
